src/http_tracker_connection.cpp

```diff
@@ -121,7 +121,7 @@ namespace libtorrent {
 				, escape_string({tracker_req().pid.data(), 20}).c_str()
 				// the i2p tracker seems to verify that the port is not 0,
 				// even though it ignores it otherwise
-				, i2p ? 1 : tracker_req().listen_port
```

Comment:
is there logic elsewhere that makes sure that `listen_port` is always zero for i2p torrents?
It's not obvious that this change is correct.

Reply:
it seems simpler to remove the condition here, and make sure we set `listen_port` to 1, wherever we set it.

Reply:
Hello, @arvidn 

> is there logic elsewhere that makes sure that listen_port is always zero for i2p torrents?

If it wasn't the case, you wouldn't have to ensure port is non-zero in the first place.

> it seems simpler to remove the condition here, and make sure we set listen_port to 1, wherever we set it.

I'm not sure about that. port=1 is a valid port and hypothetically we can have real listen_port equal to 1. On the other hand, announces is the only place where port number has to be "spoofed". We need that change only here (as i2p did before).

Reply:
> If it wasn't the case, you wouldn't have to ensure port is non-zero in the first place.

The existing code doesn't *just* make sure the port is non-zero, it also makes sure that whatever value it might have is not announced. i.e. it protects against information leaks.

I'm thinking it would likely be an overall simplification if the code building the `tracker_request` object would set the listen port to 1, and not have any special case here.

For instance, it would make UDP trackers behave the same.

> I'm not sure about that. port=1 is a valid port and hypothetically we can have real listen_port equal to 1. On the other hand, announces is the only place where port number has to be "spoofed". We need that change only here (as i2p did before).

I'm not suggesting making the *actual* listen port 1, but to fill in the `tracker_request` object with a listen port of 1, when it is supposed to not be there.

Btw, do you know what other popular clients do in this case? uTorrent, Transmission or rTorrent?

Reply:
> The existing code doesn't just make sure the port is non-zero, it also makes sure that whatever value it might have is not announced. i.e. it protects against information leaks.

I'm not familiar with existing code base that well, so I'm lost about how to implement it properly. But I don't think port number is a much of a leak as far as real client IP address is hidden. I was pretty sure `tracker_request.listen_port` tends to be `0` when proxy-only operation of client is enabled, so I had thought it is sufficient criteria and main goal is to make port number in announce look always viable.

> Btw, do you know what other popular clients do in this case? uTorrent, Transmission or rTorrent?

Transmission has no proxy support at all (it had, but proxy support was eliminated long time ago). I haven't tried uTorrent for a long time because it lost credibility long time ago when they bundled application with a background cryptocurrency miner.

qBittorrent is the client I use. It relies on libtorrent-rasterbar, so I had to make this patch to make qBittorrent actually work behind proxy. Otherwise, it published `port=0` and such announces were mostly rejected.

Also I've tried Deluge which is also based on `python-libtorrent` which inherently is a binding to libtorrent-rasterbar library. And it had similar issue: `port=0`, announces rejected, downloads not happening.

I don't know any more torrent clients with viable proxy support. Maybe rtorrent also does, but I haven't tried it yet, last time I've checked it appeared to me too complex. Probably I should revisit it and give it another chance.

BTW, qbittorrent has issues with proxy on its own: qbittorrent/qBittorrent#12114 Latest versions of qbittorrent have terrible IP address leak with older versions of libtorrent-rasterbar (pre-1.2.0).